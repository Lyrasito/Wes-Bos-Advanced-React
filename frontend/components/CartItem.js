import { Query } from "react-apollo";
import gql from "graphql-tag";
import formatMoney from "../lib/formatMoney";
import styled from "styled-components";

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${(props) => props.theme.lightGrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CART_ITEM_QUERY = gql`
  query CART_ITEM_QUERY($id: ID!) {
    cartItem(id: $id) {
      id
      quantity
      item {
        id
        title
      }
    }
  }
`;

const CartItem = (props) => {
  const { item } = props.cartItem;
  return (
    <CartItemStyles>
      <img src={item.image} width="100" alt={item.title} />
      <div className="cart-item-details">
        <h3>{item.title}</h3>
        <p>
          {formatMoney(item.price * props.cartItem.quantity)} -{" "}
          <em>
            {props.cartItem.quantity} &times; {formatMoney(item.price)} each
          </em>
        </p>
      </div>
    </CartItemStyles>
  );
};
export default CartItem;
