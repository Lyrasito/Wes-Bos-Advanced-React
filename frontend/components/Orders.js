import { Query } from "react-apollo";
import gql from "graphql-tag";
import { formatDistance } from "date-fns";
import Link from "next/link";
import styled from "styled-components";
import formatMoney from "../lib/formatMoney";
import OrderItemStyles from "./styles/OrderItemStyles";

const ALL_ORDERS_QUERY = gql`
  query ALL_ORDERS_QUERY {
    orders {
      id
      total
      createdAt
      items {
        id
        title
        description
        quantity
        price
        image
      }
    }
  }
`;

const OrdersUl = styled.ul`
  display: grid;
  grid-gap: 4rem;
  grid-template-columns: repeat(auto-fit, minmax(40%, 1fr));
`;

const Orders = (props) => {
  return (
    <Query query={ALL_ORDERS_QUERY}>
      {({ data, loading, error }) => {
        console.log(data);

        return (
          <div>
            <h2>You have {data.orders.length} orders.</h2>
            <OrdersUl>
              {data.orders.map((order) => {
                console.log(order.items);
                const itemsCount = order.items.reduce(
                  (a, b) => a + b.quantity,
                  0
                );
                console.log(itemsCount);
                return (
                  <OrderItemStyles key={order.id}>
                    <Link href={{ pathname: "/order", query: order.id }}>
                      <a>
                        <div className="order-meta">
                          <p>
                            {itemsCount} Item{itemsCount === 1 ? "" : "s"}
                          </p>
                          <p>
                            {order.items.length} Product
                            {order.items.length === 1 ? "" : "s"}
                          </p>
                          <p>{formatDistance(order.createdAt, new Date())}</p>
                          <p>{formatMoney(order.total)} </p>
                        </div>
                        <div className="images">
                          {order.items.map((item) => (
                            <img
                              key={item.id}
                              src={item.image}
                              alt={item.title}
                            />
                          ))}
                        </div>
                      </a>
                    </Link>
                  </OrderItemStyles>
                );
              })}
            </OrdersUl>
          </div>
        );
      }}
    </Query>
  );
};

export default Orders;
