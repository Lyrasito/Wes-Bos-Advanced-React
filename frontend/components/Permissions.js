import React from "react";
import { Query, Mutation } from "react-apollo";
import Error from "./ErrorMessage";
import gql from "graphql-tag";
import PropTypes from "prop-types";
import Table from "./styles/Table";
import SickButton from "./styles/SickButton";

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    users {
      name
      email
      id
      permissions
    }
  }
`;

const UPDATE_PERMISSIONS_MUTATION = gql`
  mutation UPDATE_PERMISSIONS_MUTATION(
    $userId: ID!
    $permissions: [Permission]!
  ) {
    updatePermissions(userId: $userId, permissions: $permissions) {
      name
      email
      id
      permissions
    }
  }
`;

const permissionsArray = [
  "ADMIN",
  "USER",
  "ITEMCREATE",
  "ITEMUPDATE",
  "ITEMDELETE",
  "PERMISSIONUPDATE",
];
const Permissions = (props) => (
  <Query query={ALL_USERS_QUERY}>
    {({ data, loading, error }) => {
      if (error) return <Error error={error} />;
      return (
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {permissionsArray.map((permission) => (
                  <th key={permission}>{permission}</th>
                ))}
                <th></th>
              </tr>
            </thead>
            <tbody>
              {data.users.map((user) => (
                <UserPermissions user={user} key={user.id} />
              ))}
            </tbody>
          </Table>
        </div>
      );
    }}
  </Query>
);

class UserPermissions extends React.Component {
  static propTypes = {
    user: PropTypes.shape({
      name: PropTypes.string,
      email: PropTypes.string,
      id: PropTypes.string,
      permissions: PropTypes.array,
    }).isRequired,
  };
  state = {
    permissions: this.props.user.permissions,
  };
  onChange = (e) => {
    let newPermissions = [...this.state.permissions];
    if (e.target.checked) {
      newPermissions.push(e.target.value);
    } else {
      newPermissions = newPermissions.filter(
        (permission) => permission !== e.target.value
      );
    }

    this.setState({ permissions: newPermissions });
  };
  render() {
    const user = this.props.user;
    return (
      <Mutation
        mutation={UPDATE_PERMISSIONS_MUTATION}
        variables={{ userId: user.id, permissions: this.state.permissions }}
      >
        {(updatePermissions, { called, data, loading, error }) => {
          //console.log("user", user.id, user.name);

          return (
            <>
              {error && (
                <tr>
                  <td colSpan="8">
                    <Error error={error} />
                  </td>
                </tr>
              )}
              <tr>
                <td>{user.name}</td>
                <td>{user.email}</td>
                {permissionsArray.map((permission) => (
                  <td key={permission}>
                    <label htmlFor={`${user.id}-permission-${permission}`}>
                      <input
                        type="checkbox"
                        id={`${user.id}-permission-${permission}`}
                        checked={this.state.permissions.includes(permission)}
                        value={permission}
                        onChange={this.onChange}
                      />
                    </label>
                  </td>
                ))}
                <td>
                  <SickButton onClick={updatePermissions} disabled={loading}>
                    Updat{loading ? "ing" : "e"}
                  </SickButton>
                </td>
              </tr>
            </>
          );
        }}
      </Mutation>
    );
  }
}

export default Permissions;
