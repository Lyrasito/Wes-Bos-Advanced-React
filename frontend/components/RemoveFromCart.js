import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import PropTypes from "prop-types";

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`;

const RemoveFromCart = (props) => (
  <Mutation mutation={REMOVE_FROM_CART_MUTATION} variables={{ id: props.id }}>
    {(removeFromCart, { data, loading, error }) => {
      console.log(props.id);
      console.log(data);
      console.log(error);
      return <button onClick={removeFromCart}>Remove from cart</button>;
    }}
  </Mutation>
);

export default RemoveFromCart;
