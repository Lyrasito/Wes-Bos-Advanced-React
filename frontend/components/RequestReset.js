import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./User";

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    requestReset(email: $email) {
      message
    }
  }
`;

class RequestReset extends React.Component {
  state = {
    email: "",
  };
  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };
  render() {
    return (
      <Mutation mutation={REQUEST_RESET_MUTATION} variables={this.state}>
        {(reset, { loading, error, called }) => {
          return (
            <Form
              onSubmit={async (e) => {
                e.preventDefault();
                const res = await reset();
                this.setState({ email: "" });
              }}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Request a password reset</h2>
                <Error error={error}></Error>
                {!error && !loading && called && (
                  <p>
                    Request sent successfully! Please check your email for a
                    reset link.
                  </p>
                )}
                <label htmlFor="email">
                  Email{" "}
                  <input
                    type="email"
                    name="email"
                    value={this.state.email}
                    placeholder="email"
                    onChange={this.handleChange}
                  />
                </label>

                <button type="submit">Send Request</button>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

export default RequestReset;
