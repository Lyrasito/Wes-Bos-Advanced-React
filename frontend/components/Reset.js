import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import propTypes from "prop-types";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./User";

const RESET_MUTATION = gql`
  mutation RESET_MUTATION(
    $resetToken: String!
    $password: String!
    $confirmPassword: String!
  ) {
    resetPassword(
      resetToken: $resetToken
      password: $password
      confirmPassword: $confirmPassword
    ) {
      name
      id
      email
    }
  }
`;

class Reset extends React.Component {
  static propTypes = {
    resetToken: propTypes.string.isRequired,
  };
  state = {
    password: "",
    confirmPassword: "",
  };
  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };
  render() {
    return (
      <div>
        <Mutation
          mutation={RESET_MUTATION}
          variables={{
            resetToken: this.props.resetToken,
            password: this.state.password,
            confirmPassword: this.state.confirmPassword,
          }}
          refetchQueries={[{ query: CURRENT_USER_QUERY }]}
        >
          {(reset, { loading, error, called }) => {
            return (
              <Form
                onSubmit={async (e) => {
                  e.preventDefault();
                  const res = await reset();
                  this.setState({ password: "", confirmPassword: "" });
                }}
              >
                <fieldset disabled={loading} aria-busy={loading}>
                  <h2>Enter your new password.</h2>
                  <Error error={error}></Error>
                  {!error && !loading && called && (
                    <p>Password changed successfully!</p>
                  )}
                  <label htmlFor="password">
                    Password{" "}
                    <input
                      type="password"
                      name="password"
                      value={this.state.password}
                      placeholder="password"
                      onChange={this.handleChange}
                    />
                  </label>
                  <label htmlFor="confirmPassword">
                    Confirm Password{" "}
                    <input
                      type="password"
                      name="confirmPassword"
                      value={this.state.confirmPassword}
                      placeholder="confirm password"
                      onChange={this.handleChange}
                    />
                  </label>

                  <button type="submit">Send </button>
                </fieldset>
              </Form>
            );
          }}
        </Mutation>
      </div>
    );
  }
}

export default Reset;
