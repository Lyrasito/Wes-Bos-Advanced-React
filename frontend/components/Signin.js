import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Form from "./styles/Form";
import Error from "./ErrorMessage";
import { CURRENT_USER_QUERY } from "./User";

const SIGNIN_MUTATION = gql`
  mutation SIGNIN_MUTATION($email: String!, $password: String!) {
    signin(email: $email, password: $password) {
      id
      email
      name
      password
      permissions
    }
  }
`;

class Signin extends React.Component {
  state = {
    email: "",
    password: "",
  };
  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };
  render() {
    return (
      <div>
        <Mutation
          mutation={SIGNIN_MUTATION}
          variables={this.state}
          refetchQueries={[{ query: CURRENT_USER_QUERY }]}
        >
          {(signin, { loading, error }) => {
            return (
              <Form
                onSubmit={async (e) => {
                  e.preventDefault();
                  const res = await signin();
                  this.setState({ email: "", password: "" });
                }}
              >
                <fieldset disabled={loading} aria-busy={loading}>
                  <h2>Sign in to your account</h2>
                  <Error error={error}></Error>
                  <label htmlFor="email">
                    Email{" "}
                    <input
                      type="email"
                      name="email"
                      value={this.state.email}
                      placeholder="email"
                      onChange={this.handleChange}
                    />
                  </label>
                  <label htmlFor="password">
                    Password{" "}
                    <input
                      type="password"
                      name="password"
                      value={this.state.password}
                      placeholder="password"
                      onChange={this.handleChange}
                    />
                  </label>
                  <button type="submit">Sign In</button>
                </fieldset>
              </Form>
            );
          }}
        </Mutation>
      </div>
    );
  }
}

export default Signin;
