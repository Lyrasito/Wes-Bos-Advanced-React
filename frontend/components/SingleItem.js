import React from "react";
import { Query } from "react-apollo";
import gql from "graphql-tag";
import styled from "styled-components";
import Head from "next/head";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";
import AddToCart from "./AddToCart";
import SickButton from "./styles/SickButton";

const SingleItemStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${(props) => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
  button {
    background: red;
    color: white;
    font-weight: 500;
    border: 0;
    border-radius: 0;
    text-transform: uppercase;
    font-size: 2rem;
    padding: 0.8rem 1.5rem;
    transform: skew(-2deg);
    display: inline-block;
    transition: all 0.5s;
    &[disabled] {
      opacity: 0.5;
    }
  }
`;

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      price
      largeImage
    }
  }
`;

class SingleItem extends React.Component {
  render() {
    return (
      <div>
        <Query query={SINGLE_ITEM_QUERY} variables={{ id: this.props.id }}>
          {({ loading, error, called, data }) => {
            const item = data.item;
            if (error) {
              return <Error error={error} />;
            }
            if (loading) {
              return <p>Loading...</p>;
            }
            if (!item) {
              return <p>No item found for id {this.props.id}</p>;
            }
            return (
              <SingleItemStyles>
                <Head>
                  <title>Sick Fits | {item.title}</title>
                </Head>
                <img src={item.largeImage} alt={item.title} />
                <div className="details">
                  <h2>Viewing {item.title}</h2>
                  <p>{item.description}</p>
                  <p>{formatMoney(item.price)}</p>
                  <AddToCart id={item.id} />
                </div>
              </SingleItemStyles>
            );
          }}
        </Query>
      </div>
    );
  }
}

export default SingleItem;
