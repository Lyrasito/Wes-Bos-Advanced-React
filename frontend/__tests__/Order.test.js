import { mount } from "enzyme";
import wait from "waait";
import { ApolloConsumer } from "react-apollo";
import NProgress from "nprogress";
import Router from "next/router";
import OrderComponent, { SINGLE_ORDER_QUERY } from "../components/Order";
import toJSON from "enzyme-to-json";
import { MockedProvider } from "react-apollo/test-utils";
import { CURRENT_USER_QUERY } from "../components/User";
import { fakeUser, fakeCartItem, fakeOrder } from "../lib/testUtils";

const mocks = [
  {
    request: {
      query: CURRENT_USER_QUERY,
    },
    result: {
      me: {
        ...fakeUser(),
        orders: [fakeOrder({ id: "singleorderid" })],
      },
    },
  },
  {
    request: {
      query: SINGLE_ORDER_QUERY,
      variables: { id: "singleorderid" },
    },
    result: {
      data: {
        order: {
          ...fakeOrder(),
          id: "singleorderid",
        },
      },
    },
  },
];

describe("<OrderComponent />", () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <OrderComponent id="singleorderid" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const order = wrapper.find("div[data-test='order']");
    expect(toJSON(order)).toMatchSnapshot();
  });
});
