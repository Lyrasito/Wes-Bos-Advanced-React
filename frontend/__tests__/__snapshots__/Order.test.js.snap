// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderComponent /> renders and matches snapshot 1`] = `
<div
  className="OrderStyles-sc-4oqalm-0 camMij"
  data-test="order"
>
  <SideEffect(Head)>
    <Head />
  </SideEffect(Head)>
  <p>
    <span>
      Order Id:
    </span>
    <span>
      singleorderid
    </span>
  </p>
  <p>
    <span>
      Charge:
    </span>
    <span>
      ch_123
    </span>
  </p>
  <p>
    <span>
      Date:
    </span>
    <span>
      March 31, 2018 8:00 PM
    </span>
  </p>
  <p>
    <span>
      Total:
    </span>
    <span>
      $400
    </span>
  </p>
  <div
    className="items"
  >
    <div
      className="order-item"
      key="77ee1755-9f56-4760-816b-dd266da805ab"
    >
      <img
        alt="explicabo eaque necessitatibus sit suscipit omnis necessitatibus"
        src="doloremque.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          explicabo eaque necessitatibus sit suscipit omnis necessitatibus
        </h2>
        <p>
          Qty: 
          1
        </p>
        <p>
          Each: 
          $42.34
        </p>
        <p>
          Subtotal: 
          $42.34
        </p>
        <p>
          maiores ullam quae corporis qui aut nobis
        </p>
      </div>
    </div>
    <div
      className="order-item"
      key="ccad4806-718d-401c-bd82-c4e59edfec40"
    >
      <img
        alt="sapiente nostrum aut corrupti sed consectetur soluta"
        src="iusto.jpg"
      />
      <div
        className="item-details"
      >
        <h2>
          sapiente nostrum aut corrupti sed consectetur soluta
        </h2>
        <p>
          Qty: 
          1
        </p>
        <p>
          Each: 
          $42.34
        </p>
        <p>
          Subtotal: 
          $42.34
        </p>
        <p>
          inventore laudantium dolor sint officiis qui dignissimos
        </p>
      </div>
    </div>
  </div>
</div>
`;
